File: pydot-rs/dot_python/parser.py
from collections import defaultdict
from enum import Enum, auto
from typing import Iterable

from .tokenizer import Token, BasicToken, IdentifierToken
from .graph import Graph


class ParserState(Enum):
    Start = auto()
    ExpectGraphName = auto()
    ExpectLBracket = auto()
    ExpectNodeName = auto()
    ExpectEdgeOrSemicolon = auto()
    ExpectNodeNameOrRBracket = auto()
    End = auto()


class Parser:
    def __init__(self):
        self.state = ParserState.Start
        self.node_map: dict[str, int] = {}
        self.current_chain: list[int] = []

        self.graph_name: str | None = None
        self.nodes: list[str] = []
        self.adjacency: defaultdict[int, list[int]] = defaultdict(list)

    def get_node_index(self, name: str) -> int:
        """Get the index of a node, creating it if it doesn't exist"""
        node_idx = self.node_map.get(name, None)
        if node_idx is not None:
            return node_idx
        node_idx = len(self.nodes)
        self.nodes.append(name)
        return node_idx

    def persist_current_chain(self):
        """Persist the current chain of nodes as a path in the graph"""
        for i in range(len(self.current_chain) - 1):
            self.adjacency[self.current_chain[i]].append(self.current_chain[i + 1])
        self.current_chain = []

    def parse(self, tokens: Iterable[Token]) -> Graph:
        """Parse a list of tokens into a graph instance"""
        token_iter = iter(tokens)
        while self.state != ParserState.End:
            token = next(token_iter)
            self.parse_token(token)
        return Graph(
            graph_name=self.graph_name,
            nodes=self.nodes,
            adjacency=dict(self.adjacency),
        )

    def parse_token(self, token: Token):
        """Parse a single token"""
        match (self.state, token):
            # Graph header (type, name, bracket)
            case (ParserState.Start, BasicToken.GRAPH):
                self.state = ParserState.ExpectGraphName
            case (ParserState.ExpectGraphName, IdentifierToken(name)):
                self.graph_name = name
                self.state = ParserState.ExpectLBracket
            case (ParserState.ExpectLBracket, BasicToken.LEFT_BRACKET):
                self.state = ParserState.ExpectNodeNameOrRBracket

            # Graph body (nodes and edges)
            case (
                ParserState.ExpectNodeNameOrRBracket | ParserState.ExpectNodeName,
                IdentifierToken(name),
            ):
                node_idx = self.get_node_index(name)
                self.current_chain.append(node_idx)
                self.state = ParserState.ExpectEdgeOrSemicolon

            case (
                ParserState.ExpectEdgeOrSemicolon,
                BasicToken.EDGE,
            ):
                self.state = ParserState.ExpectNodeName

            case (ParserState.ExpectEdgeOrSemicolon, BasicToken.SEMICOLON):
                self.persist_current_chain()
                self.state = ParserState.ExpectNodeNameOrRBracket

            # Graph end (right bracket)
            case (ParserState.ExpectNodeNameOrRBracket, BasicToken.RIGHT_BRACKET):
                self.state = ParserState.End

            # Error cases (unexpected tokens)
            case _:
                raise Exception(f"Unexpected token {token} in state {self.state}")

File: pydot-rs/dot_rust/src/parser.rs
use std::collections::HashMap;

use crate::{tokenizer::Token, Graph};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum ParserState {
    Start,
    ExpectGraphName,
    ExpectLBracket,
    ExpectNodeName,
    ExpectEdgeOrSemicolon,
    ExpectNodeNameOrRBracket,
    End,
}

pub struct Parser {
    state: ParserState,
    node_map: HashMap<String, usize>,
    current_chain: Vec<usize>,

    graph_name: String,
    nodes: Vec<String>,
    adjacency: HashMap<usize, Vec<usize>>,
}

impl Parser {
    pub fn new() -> Self {
        Self {
            state: ParserState::Start,
            node_map: HashMap::new(),
            current_chain: Vec::new(),

            graph_name: String::new(),
            nodes: Vec::new(),
            adjacency: HashMap::new(),
        }
    }

    /// Get the index of a node, or insert it if it doesn't exist.
    fn get_node_index(&mut self, name: &str) -> usize {
        *self.node_map.entry(name.to_string()).or_insert_with(|| {
            let index = self.nodes.len();
            self.nodes.push(name.to_string());
            index
        })
    }

    /// Add an edge between two nodes, in both directions
    fn add_undirected_edge(&mut self, a: usize, b: usize) {
        self.adjacency.entry(a).or_insert_with(Vec::new).push(b);
        self.adjacency.entry(b).or_insert_with(Vec::new).push(a);
    }

    /// Persist the current chain of nodes as a path in the graph
    fn persist_current_chain(&mut self) {
        for i in 0..self.current_chain.len() - 1 {
            self.add_undirected_edge(self.current_chain[i], self.current_chain[i + 1]);
        }
        self.current_chain.clear();
    }

    pub fn parse(&mut self, tokens: &mut impl Iterator<Item = Token>) -> Graph {
        while self.state != ParserState::End {
            let token = tokens.next().unwrap();
            self.parse_token(token);
        }
        Graph {
            graph_name: self.graph_name.clone(),
            nodes: self.nodes.clone(),
            adjacency: self.adjacency.clone(),
        }
    }

    /// Parse a single token
    fn parse_token(&mut self, token: Token) {
        match (self.state, token) {
            // Graph header (type, name, bracket)
            (ParserState::Start, Token::Graph) => {
                self.state = ParserState::ExpectGraphName;
            }
            (ParserState::ExpectGraphName, Token::Identifier(name)) => {
                self.graph_name = name;
                self.state = ParserState::ExpectLBracket;
            }
            (ParserState::ExpectLBracket, Token::LeftBracket) => {
                self.state = ParserState::ExpectNodeName;
            }

            // Graph Body
            (
                ParserState::ExpectNodeNameOrRBracket | ParserState::ExpectNodeName,
                Token::Identifier(name),
            ) => {
                let index = self.get_node_index(&name);
                self.current_chain.push(index);
                self.state = ParserState::ExpectEdgeOrSemicolon;
            }
            (ParserState::ExpectEdgeOrSemicolon, Token::Edge) => {
                self.state = ParserState::ExpectNodeName;
            }
            (ParserState::ExpectEdgeOrSemicolon, Token::Semicolon) => {
                self.persist_current_chain();
                self.state = ParserState::ExpectNodeNameOrRBracket;
            }

            // Graph End
            (ParserState::ExpectNodeNameOrRBracket, Token::RightBracket) => {
                self.state = ParserState::End;
            }

            // Error
            (state, token) => {
                panic!("Unexpected token {:?} in state {:?}", token, state);
            }
        }
    }
}
